File: backend/src/middlewares/errorhandler.ts
import { ErrorRequestHandler, NextFunction, Request, Response } from 'express';
import { CustomAPIError } from '../errors/custom-api.js';
import { StatusCodes } from 'http-status-codes';

export const errorHandlerMiddleware: ErrorRequestHandler = (
  err,
  req: Request,
  res: Response,
  next: NextFunction,
) => {
  if (process.env.NODE_ENV === 'development') {
    console.error(err);
  }
  let customError = {
    statusCode: StatusCodes.INTERNAL_SERVER_ERROR,
    msg: 'Something went wrong try again later',
  };
  if (err instanceof CustomAPIError) {
    customError.statusCode = err.statusCode;
    customError.msg = err.message;
  }

  return res.status(customError.statusCode).json({ msg: customError.msg });
};

File: backend-go/internal/httpapi/errorhandler.go
package httpapi

import (
	"errors"
	"log"
	"net/http"

	"github.com/gofiber/fiber/v2"

	"ratracer/backend/internal/apierror"
	"ratracer/backend/internal/config"
)

func ErrorHandler(cfg *config.Config) fiber.ErrorHandler {
	return func(c *fiber.Ctx, err error) error {
		if !cfg.IsProduction() {
			log.Println(err)
		}

		statusCode := http.StatusInternalServerError
		msg := "Something went wrong try again later"

		var apiErr *apierror.APIError
		var fiberErr *fiber.Error
		switch {
		case errors.As(err, &apiErr):
			statusCode = apiErr.StatusCode
			msg = apiErr.Message
		case errors.As(err, &fiberErr):
			statusCode = fiberErr.Code
			msg = fiberErr.Message
		}

		return c.Status(statusCode).JSON(fiber.Map{"msg": msg})
	}
}
